Reduce logging related IPC traffic

pvollan opened the pull request:
#### b511c5d5b5a7adf0a26c41907f5ab46f4b0fcaac
<pre>
Reduce logging related IPC traffic
<a href="https://bugs.webkit.org/show_bug.cgi?id=255787">https://bugs.webkit.org/show_bug.cgi?id=255787</a>
rdar://108369535

Reviewed by NOBODY (OOPS!).

Reduce logging related IPC traffic by sending batches of logs.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::logOnBehalfOfWebContent):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::WebPaymentCoordinatorProxy::platformPaymentRequest):
* Source/WebKit/Shared/Cocoa/LogData.h: Added.
* Source/WebKit/Shared/Cocoa/LogData.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::m_logFlushTimer):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::registerLogHook):
(WebKit::WebProcess::logFlushTimerFired):
(WebKit::registerLogHook): Deleted.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b511c5d5b5a7adf0a26c41907f5ab46f4b0fcaac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4146 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4258 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4376 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5608 "Failed to compile WebKit") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4390 "Failed to compile WebKit") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4363 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4223 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/5608 "Failed to compile WebKit") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4204 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4390 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3741 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5601 "Failed to compile WebKit") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1886 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3718 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/5601 "Failed to compile WebKit") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3713 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3782 "1 flakes 3 failures") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/5601 "Failed to compile WebKit") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4181 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3388 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3695 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3716 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/7823 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3975 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1415,17 +1416,34 @@ void NetworkConnectionToWebProcess::installMockContentFilter(WebCore::MockConten
+            msg.privdata = log.privdata.data();
+            msg.privdata_sz = log.privdata.size();
+            
+            char* messageString = os_log_copy_message_string(&msg);
```
You can use MallocPtr here, to avoid needing a manual call to free().

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1415,17 +1416,34 @@ void NetworkConnectionToWebProcess::installMockContentFilter(WebCore::MockConten
+{
+    static NeverDestroyed<Ref<WorkQueue>> queue(WorkQueue::create("Log Queue", WorkQueue::QOS::Background));
+
+    queue.get()->dispatch([logs = WTFMove(logs), pid = pid, webProcessIdentifier = m_webProcessIdentifier] {
```
`= pid` is unnecessary.

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1415,17 +1416,34 @@ void NetworkConnectionToWebProcess::installMockContentFilter(WebCore::MockConten
-        osLogChannel = adoptOSObject(os_log_create(channel, category));
-    }
-    RELEASE_ASSERT(logType == OS_LOG_TYPE_DEFAULT || logType == OS_LOG_TYPE_INFO || logType == OS_LOG_TYPE_DEBUG || logType == OS_LOG_TYPE_ERROR || logType == OS_LOG_TYPE_FAULT);
+void NetworkConnectionToWebProcess::logOnBehalfOfWebContent(Vector<LogData>&& logs, int32_t pid)
```
Would be nice to use `ProcessID` from WTF for the pid.

geoffreygaren commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -663,35 +663,25 @@ static void registerLogHook()
-                IPC::Connection::send(connectionID, Messages::NetworkConnectionToWebProcess::LogOnBehalfOfWebContent(logChannel.bytesInludingNullTerminator(), logCategory.bytesInludingNullTerminator(), logString, type, getpid()), 0);
+        m_logSet.append(WTFMove(logItem));
+        if (m_logSet.size() < 32)
+            return;
```
We probably need a timer to to eventually send the log set even if the counter never reaches 32. Otherwise, the final 31 messages in any process will be lost.

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1415,17 +1416,34 @@ void NetworkConnectionToWebProcess::installMockContentFilter(WebCore::MockConten
-    RELEASE_ASSERT(logType == OS_LOG_TYPE_DEFAULT || logType == OS_LOG_TYPE_INFO || logType == OS_LOG_TYPE_DEBUG || logType == OS_LOG_TYPE_ERROR || logType == OS_LOG_TYPE_FAULT);
+void NetworkConnectionToWebProcess::logOnBehalfOfWebContent(Vector<LogData>&& logs, int32_t pid)
+{
+    static NeverDestroyed<Ref<WorkQueue>> queue(WorkQueue::create("Log Queue", WorkQueue::QOS::Background));
```
This should be using a WorkQueueMessageReceiver. It seems unfortunate to send those log lines to the main thread only to re-dispatch them to a background queue.

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1415,17 +1416,34 @@ void NetworkConnectionToWebProcess::installMockContentFilter(WebCore::MockConten
+            
+            char* messageString = os_log_copy_message_string(&msg);
+            if (!messageString)
+                return;
```
`continue;` not return.

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1415,17 +1416,34 @@ void NetworkConnectionToWebProcess::installMockContentFilter(WebCore::MockConten
+            OSObjectPtr<os_log_t> osLogChannel;
+            if (log.channel.data() && log.category.data())
+                osLogChannel = adoptOSObject(os_log_create(log.channel.data(), log.category.data()));
+            RELEASE_ASSERT(log.type == OS_LOG_TYPE_DEFAULT || log.type == OS_LOG_TYPE_INFO || log.type == OS_LOG_TYPE_DEBUG || log.type == OS_LOG_TYPE_ERROR || log.type == OS_LOG_TYPE_FAULT);
```
Shouldn't this be a MESSAGE_CHECK? It's not OK for bad IPC from a compromised Webprocess to kill our shared network process.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -186,7 +186,7 @@
 #if __has_include(<WebKitAdditions/InternalBuildAdditions.h>)
         if (isInternalBuild())
-            disableLogging = false;
+            disableLogging = true;
```
I don't really understand this change.

cdumez commented on Source/WebKit/WebProcess/WebProcess.h:
```diff
@@ -771,6 +779,11 @@ class WebProcess : public AuxiliaryProcess
+    
+#if ENABLE(LOGD_BLOCKING_IN_WEBCONTENT)
+    Lock m_logDataLock;
+    Vector<LogData> m_logSet WTF_GUARDED_BY_LOCK(m_logDataLock);
```
We probably don't want to use "Set" in the name. I find it a little confusing for a vector.

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -663,35 +663,25 @@ static void registerLogHook()
-            if (connectionID)
-                IPC::Connection::send(connectionID, Messages::NetworkConnectionToWebProcess::LogOnBehalfOfWebContent(logChannel.bytesInludingNullTerminator(), logCategory.bytesInludingNullTerminator(), logString, type, getpid()), 0);
+        m_logSet.append(WTFMove(logItem));
+        if (m_logSet.size() < 32)
```
32 should be a constant with a nice name.

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -663,35 +663,25 @@ static void registerLogHook()
+        auto connectionID = networkProcessConnectionID();
+        if (!connectionID)
+            return;
+        IPC::Connection::send(connectionID, Messages::NetworkConnectionToWebProcess::LogOnBehalfOfWebContent(WTFMove(m_logSet), getpid()), 0);
```
Should be using std::exchange(), not WTFMove() given that you'll reuse m_logSet later on. WTFMove() is just a cast. This code is fragile because it both:
- Relies on the Vector move constructor to leave the Vector in a reusable & empty state
- Relies on LogOnBehalfOfWebContent to actually adopt the Vector.

cdumez commented on Source/WebKit/Shared/Cocoa/LogData.h:
```diff
@@ -0,0 +1,93 @@
+};
+
+template<class Encoder>
+void LogData::encode(Encoder& encoder) const
```
We don't write manual serializers / deserializer like this anymore. Please use the new format in a serialization.in.

pvollan replied:
Thanks for reviewing, Geoff and Chris!

These are all very good points, I will update the patch accordingly.

pvollan committed:
Reduce logging related IPC traffic
https://bugs.webkit.org/show_bug.cgi?id=255787
rdar://108369535

Reviewed by NOBODY (OOPS!).

Reduce logging related IPC traffic by sending batches of logs.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::logOnBehalfOfWebContent):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::WebPaymentCoordinatorProxy::platformPaymentRequest):
* Source/WebKit/Shared/Cocoa/LogData.h: Added.
* Source/WebKit/Shared/Cocoa/LogData.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::m_logFlushTimer):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::registerLogHook):
(WebKit::WebProcess::logFlushTimerFired):
(WebKit::registerLogHook): Deleted.
